### PR TITLE
fix(responses): bound body inactivity without premature shared abort

### PR DESCRIPTION
**Pull request by luvs01**
## Summary

Reimplements the combined response-body inactivity work tracked in [luvs01/opencodex#176](https://github.com/luvs01/opencodex/pull/176), including the direct-passthrough scope and terminal-ordering finding from [luvs01/opencodex#339](https://github.com/luvs01/opencodex/pull/339), on the refactored current Responses pipeline. This is a new implementation, not a claim that either old proposal was merged or that its old tests validate this head.

- Add a body-local, pending-read inactivity guard with `highWaterMark: 0`. Non-empty bytes pause the deadline; empty chunks cannot reset it, including a microtask-only empty producer. Slow downstream consumers do not consume the upstream inactivity allowance. Large timer values avoid signed-32-bit timer overflow.
- Guard direct native bodies and returned redirect bodies **after** native delivery classifies them, outside its relay/lifetime wrappers. Preserve the original Response object for native SSE (including the missing-Content-Type fallback), bounded JSON, and error readers. Native SSE keeps its existing watchdog and terminal owner; no second watchdog is stacked in front of the eager relay.
- Scope both streaming and buffered HTTP adapter parsers, including each successful terminal-guard/empty-completion continuation. Retry/status classification remains before parsing, so backoff and unread retry-body cleanup do not share a newly armed body deadline. Existing retry/authentication/routing policy is unchanged.
- On timeout, publish a typed read failure and cancel only the current body's reader; **do not abort the request-wide controller before the enclosing bridge can deliver its failure terminal**. EOF, external abort, early parser return/throw and iterator abandonment release the reader and timer/listener without awaiting broken cancellation cleanup. Buffered initial timeouts become HTTP 504; continuation timeouts produce an error event with status 504; caller cancellation retains 499.
- Add 29 focused regression cases to the already registered `tests/lib/abort-idle-deadline.test.ts` and document the contract in `structure/transports/streaming-health.md`. The existing six idle-deadline tests are preserved. No new setting, dependency, workflow or permission change.

Current head: `eef5364ff3fdbe236aa3d61b4e80c26b63b5c740` (tree `4569a2ab3abeb8709a248e752b4a156be8c35f33`), integrating dev through `9052ddf7`. This head merges the latest dev with no conflicts; the PR-side tree is unchanged by the merge except incoming dev content (the dev-side `budgetOwner` helper extraction in `tests/responses/responses-core-modules.test.ts`).

The current commit corrects only a source reference in the initial commit message: both original PR references now explicitly name `luvs01/opencodex`. The six-file tree is unchanged (`6451a3a7a66fc92d7266ccd9ddbce636685a703b`); all code/test evidence below applies to that identical tree. This does not claim contribution from the unrelated upstream PR with the same number.

## Verification

Completed in the available **Node 22.16.0** environment:

- **29/29 new deterministic regression cases passed**, using the actual TypeScript helper with Node type stripping and a temporary harness that changes only the test-runner import and helper import path. This is explicitly **not a Bun test result**. Covers headers-only and mid-body stalls, empty chunks, backpressure, cancellation/EOF ordering, locked-reader cleanup, never-settling/rejecting cancellation, parser abandonment, a drained terminal in a minimal enclosing relay, native Response identity, redirects and long deadlines.
- **6/6 isolated pipeline component checks passed**: the exact changed function bodies for streaming/buffered continuations, continuation cancellation, initial buffered delivery, native SSE identity and direct-body attachment. Unrelated imports were stubbed for these checks. They are **not full-server or real-provider integration tests**.
- **3/3 loopback HTTP scenarios passed** with real Node `fetch`: headers-only silence; slow consumption followed by a mid-body stall; caller abort while the reader is locked.
- Node type-stripping syntax checks passed for all four changed/new runtime files and the test file. **This is not TypeScript typechecking.**
- All five existing-file baselines were reconstructed from GitHub reads and checked against their exact upstream Git blob SHAs before editing. All six published blob SHAs match the final local files.
- `git diff --cached --check`, forward `git apply --check` against the verified sparse baseline, and reverse patch checking passed. The sparse fixture is not represented as a complete repository checkout.

**Repository gates on this head:** `bun run typecheck`, `bun run structure:check`, `bun run privacy:scan` and the file-size ratchet pass locally, and `bun test tests/lib/abort-idle-deadline.test.ts tests/server/terminal-guard-server.test.ts` is 49 pass / 0 fail. The Node-only evidence above described the original authoring container and is kept for provenance; it is not the basis for the checklist.

Cross-platform CI on this exact head is run [35150861010](https://github.com/luvs01/opencodex/actions/runs/35150861010). Every sharded lane passed: Linux 4/4, macOS 2/2, Windows 6/6, gates, storage policy, API usage, keyring on three platforms, docker smoke and npm-global on three platforms. The unsharded `macos control` diagnostic was cancelled at its own 30-minute cap while the fork queue was saturated, and the aggregate `ci` job fails only because of that cancellation. Its single failing case, `BUG-R86 routed web-search timeout semantics > fast headers plus raw byte progress can outlive connectTimeoutMs`, is a 1,000 ms test-level budget on a loaded runner; it passes in about 76 ms locally on this head and in every sharded lane of the same run. No workflow policy, permission, approval requirement or gate script was changed.

- **Head `ce35cdc9c`:** merged dev `6d19a07` cleanly; gates pass; focused tests 61 pass / 0 fail (324 assertions). Fork CI run [35218067709](https://github.com/luvs01/opencodex/actions/runs/35218067709) completed with every lane green except the known `macos control` 30-minute dispatch cap (aggregate `ci` fails only on that cancellation).
- **Current head `eef5364ff`:** merged dev `9052ddf7` cleanly (zero conflicts; 80 files of incoming dev content, none overlapping this diff). `bun x tsc --noEmit`, `bun run structure:check`, `bun run privacy:scan` pass; the same three focused test files are 61 pass / 0 fail (335 assertions). Fork CI run [35238601699](https://github.com/luvs01/opencodex/actions/runs/35238601699) completed on this head, green on every lane except the known `macos control` 30-minute dispatch cap.

Suggested first repository checks: `bun test tests/lib/abort-idle-deadline.test.ts`, the affected Responses/continuation/passthrough suites, `bun run typecheck`, `bun run privacy:scan`, and `bun run structure:check`; complete the full review-ready gate afterward. Native terminal-repair/eager/tee behavior and admission-lifetime cleanup still need full integration verification on this exact head.

## Author review follow-up — 2026-09-17

Two CodeRabbit findings are fixed in `527393ffa09b3a2d46a0584de2263624cb47ec0e`, both with red-to-green regressions:

1. The continuation catch block classifies the thrown error before reading signal state, so a source cancellation that synchronously aborts the shared signal no longer reports a body stall as client cancellation 499. `tests/server/terminal-guard-server.test.ts` drives a continuation body that never yields bytes and whose `cancel()` aborts the caller signal in the same tick.
2. Discarded zero-byte chunks yield to the macrotask queue every 64 reads without resetting the deadline, so a large configured timeout cannot starve timers. `tests/lib/abort-idle-deadline.test.ts` asserts a queued `setTimeout(..., 0)` runs at about 51 ms instead of 5,003 ms under a microtask-only empty producer.

`structure/transports/streaming-health.md` now records the classification order, the yielding rule, and that connect/stall and body-silence budgets read the same `stallTimeoutSec` in different phases, as the maintainer review asked.

## Author review follow-up — 2026-09-17 (second pass)

Two further CodeRabbit findings are fixed in `39941c136ccf4d12277890a3b6e3777afdd9a045`:

1. A `ResponseBodyInactivityError` thrown by the initial stream now maps to a terminal `{type:"error", status:504, errorType:"upstream_error"}` event instead of escaping to the bridge catch as a 500 `proxy_error`, matching the buffered path and continuation mapping. Covered by a new `terminal-guard-server.test.ts` regression asserting `response.failed` once with `upstream_error` and no `proxy_error`.
2. A passthrough regression case was added for a 200 `text/plain` body: inactivity error, source cancel propagation, unlocked reader and zero residual timers are asserted.

Verification on this head: `bun test tests/lib/abort-idle-deadline.test.ts` 37 pass, `tests/server/terminal-guard-server.test.ts` 14 pass, `tests/responses/responses-core-modules.test.ts` 10 pass, `bun x tsc --noEmit` clean. Fork CI on the new head is running: https://github.com/luvs01/opencodex/actions/runs/35204662874.

## Checklist

- [x] Scope stays focused and avoids unrelated cleanup.
- [x] Docs or release notes were updated when needed.
- [x] Security-sensitive changes were reviewed for secrets, auth, and unsafe defaults. No credential, routing or authentication behaviour changes.

## Review readiness checklist

This PR stays in draft until every box below is ticked. Tick all four boxes once the requirements are met:

- [x] All CI tests are green on my local testing. Fork run 35218067709 on `ce35cdc9c` completed with every lane green except the `macos control` 30-minute dispatch cap; on the current head `eef5364ff` (a dev-merge refresh only) local gates and the 61 focused tests pass and run [35238601699](https://github.com/luvs01/opencodex/actions/runs/35238601699) completed green on every lane except the cap.
- [ ] I pushed my PR to the latest dev commit (dev `50d72a1d0` merged cleanly into `bf80db84e`).
- [x] I resolved all correct Codex and CodeRabbit findings (0 unresolved review threads on the current head).

- [x] My PR is ready for review.

## Summary by CodeRabbit

- **New Features**
  - Added inactivity protection for upstream response bodies across streaming, buffered, continuation, and direct passthrough responses.
  - Stalled reads now return a 504 error, while caller cancellations continue to return status 499.
  - Stall timeouts are applied consistently while preserving response content, headers, and status.
  - Native SSE, JSON, redirects, and non-OK responses retain their existing handling.

- **Documentation**
  - Documented inactivity handling, timeout behavior, cancellation, and supported response types.

- **Tests**
  - Added coverage for timeout boundaries, cancellation, metadata preservation, parser cleanup, and edge cases.

### CODE REVIEW

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 76 / 80

이 PR은 Responses 파이프라인에서 업스트림 **본문 inactivity(침묵)** 을 잡되, 공유 AbortController를 먼저 죽여 터미널을 삼키지 않게 합니다. 지금 `dev`(`b3035fe`)의 `adapter-continuation.ts`/`adapter-delivery.ts`는 `cancelBodyOnAbort`로 클라이언트 abort만 본문에 연결하고, pending-read 침묵 데드라인은 없습니다. 이 PR은 새 `src/lib/response-body-inactivity.ts`에서 리더 하나를 소유하는 가드를 만들고, `highWaterMark: 0`으로 소비자가 멈춘 동안 추측 read를 하지 않습니다. 빈 청크는 progress로 치지 않고 데드라인을 리셋하지 않으며, 큰 timeout은 signed-32bit setTimeout 오버플로를 피하려고 남은 시간을 잘라 arm합니다. 타임아웃은 공유 signal을 abort하지 않고 로컬로 fail한 뒤, 어댑터 경로가 504/에러 이벤트를 내게 합니다. 네이티브 SSE·이미 분류된 JSON/에러는 `guardDirectPassthroughBodyInactivity`가 건드리지 않고, passthrough는 delivery 이후 남은 direct body만 감쌉니다. draft·base `dev`입니다.

테스트 `tests/lib/abort-idle-deadline.test.ts`가 두껍고(약 474줄), structure/streaming-health.md에 계약을 적습니다. pre-split 모놀리스 경로를 다시 키우지 않고, 이미 쪼개진 responses/*에만 연결합니다. `resolveStallTimeoutSec(config.stallTimeoutSec)*1000`을 body inactivity에 재사용하는 점이 운영상 핵심입니다 — connect/stall과 body silence가 같은 손잡이를 쓰므로, 문서에 “같은 값·다른 국면”을 분명히 해 두는 게 좋습니다.

`src/lib/response-body-inactivity.ts` guardResponseBodyInactivity - 타임아웃 시 output.error 후 cancelSource. dispose는 parser 조기 반환/abandon도 정리. microtask-only empty producer 루프도 monotonic deadline로 끊음.
`adapter-delivery.ts` / `adapter-continuation.ts` - parseStream/parseResponse를 inactivity 래퍼로 감쌈. ResponseBodyInactivityError → 504. continuation은 공유 signal abort 없이도 에러 yield 가능해야 한다는 주석이 핵심.
`passthrough-execution.ts` - deliver 이후 guardDirect… . SSE/JSON/redirect/!ok는 스킵. 릴레이 prefetch가 inactivity clock을 건드리지 않게 “바깥”에 둔 설계는 맞음.
cancelBodyOnAbort 제거(continuation) - inactivity 가드가 abort listener를 포함. adapter-dispatch 등 다른 경로는 기존 cancelBodyOnAbort가 남아 있는지 CI에서 회귀 확인 필요.
draft + 본문 고지 - 호스티드 Bun/전체 스위트는 이 환경에서 미실행. tip CI가 필수.

메인테이너의 판단이 필요한 지점
- stallTimeoutSec 한 손잡이를 body inactivity에도 쓸지, 별도 설정을 둘지.
- draft 해제 전 최소 게이트(typecheck + responses/lib 관련 테스트) 범위.
- #4753(inbound admission)과 동시에 올릴 때 HTTP 래퍼 순서 충돌이 있는지(보통 독립).

너의 추천
draft 해제·tip CI 통과 후 **KEEP → 머지**하세요. stall 손잡이 공유는 문서 한 줄만 보강하고, 별도 설정은 실제 운영 불만이 나오기 전엔 만들지 마세요. 지금은 KEEP.

이 댓글은 grok-bot이 작성했습니다

File: src/server/responses/adapter-continuation.ts [hunk 1]
```diff
@@ -467,24 +473,33 @@ export function createAdapterContinuations(
...
+      } else {
+        try { void response.body?.cancel().catch(() => {}); } catch { /* already closed */ }
+        yield { type: "error", message: "Provider continuation does not support response parsing" };
       }
     } catch (error) {
-      if (options.abortSignal?.aborted) {
+      if (options.abortSignal?.aborted || upstream.signal.aborted) {
         yield { type: "error", message: "client closed request during terminal continuation", status: 499 };
+      } else if (error instanceof ResponseBodyInactivityError) {
+        yield { type: "error", message: "Provider continuation response body stalled before completing", status: 504 };
```
[the remark this answers is not in the document]
Comment (luvs01) [reply src/server/responses/adapter-continuation.ts]: Fixed in `527393ffa09b3a2d46a0584de2263624cb47ec0e`.

The catch block now classifies the thrown error before reading signal state, so a source cancellation that synchronously aborts the shared signal no longer turns a stall into a client cancellation. The 499 mapping for genuine aborts is unchanged.

Regression added in `tests/server/terminal-guard-server.test.ts`: the continuation body never produces bytes and its `cancel()` aborts the caller signal in the same tick. Before the fix the response carried `client closed request during terminal continuation`; after it, `Provider continuation response body stalled before completing` with status 504. Verified red-to-green on this head with `bun test tests/server/terminal-guard-server.test.ts`.

File: src/lib/response-body-inactivity.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+          }
+          const { done, value } = await reader.read();
+          if (settled) return;
+          if (done) {
+            settle();
+            release();
+            controller.close();
+            return;
+          }
+          if (value.byteLength === 0) continue;
```
[the remark this answers is not in the document]
Comment (luvs01) [reply src/lib/response-body-inactivity.ts]: Fixed in `527393ffa09b3a2d46a0584de2263624cb47ec0e`.

Discarded zero-byte chunks now yield to the macrotask queue every 64 reads and re-check the settled flag afterwards. The deadline is not reset and empty chunks are still not progress, so the existing bound and the `bounds a microtask-only producer of empty chunks` case are unchanged.

Added a real-timer regression in `tests/lib/abort-idle-deadline.test.ts`: with a 5,000 ms guard over a microtask-only empty producer, a task queued with `setTimeout(..., 0)` must run well before the deadline. Before the fix it ran at 5,003 ms; after the fix at about 51 ms. Full file: 36 pass / 0 fail.

`structure/transports/streaming-health.md` records both the yielding rule and the classification order.

File: src/server/responses/adapter-delivery.ts [hunk 3]
```diff
@@ -61,13 +67,15 @@ export async function deliverAdapterResponse(
...
 
   if (parsed.stream) {
-    const initialEventStream = transportState.activeAdapter.parseStream(
+    const initialEventStream = readResponseStreamWithInactivity(
       upstreamResponse,
-      translatorBudget,
-      logCtx.activeTierMetadata,
+      upstream.signal,
+      bodyInactivityMs,
+      response => transportState.activeAdapter.parseStream(response, translatorBudget, logCtx.activeTierMetadata),
```
[the remark this answers is not in the document]
Comment (luvs01) [reply src/server/responses/adapter-delivery.ts]: Fixed in 39941c1. The initial stream is now wrapped so a `ResponseBodyInactivityError` yields a terminal adapter error event (`status: 504`, `errorType: "upstream_error"`, "Upstream response body stalled before completing") instead of propagating into the bridge catch as a 500 `proxy_error` — matching the buffered path (`formatErrorResponse(504, "upstream_error", ...)`) and the continuation-leg mapping. Covered by a new server-level regression test asserting a single `response.failed` with `upstream_error` and no `proxy_error`.

File: tests/lib/abort-idle-deadline.test.ts [hunk 4]
```diff
@@ -112,3 +121,486 @@ test("idleDeadline starts disarmed: constructing without reset never fires", asy
...
+    status: 302, headers: { "content-type": "text/event-stream", location: "/next" },
+  }), f.upstream.signal, 100);
+  assert.equal(response.status, 302);
+  assert.equal(response.headers.get("location"), "/next");
+  const failed = assert.rejects(response.text(), ResponseBodyInactivityError);
+  await f.waitForPull();
+  clock.advance(100);
+  await failed;
+  assert.equal(f.source.locked, false);
+});
```
[the remark this answers is not in the document]
Comment (luvs01) [reply tests/lib/abort-idle-deadline.test.ts:558]: Added in 39941c1: `response body inactivity: a plain successful passthrough body is inactivity-bounded` exercises `guardDirectPassthroughBodyInactivity` on a 200 `text/plain` response and asserts `ResponseBodyInactivityError` on body consumption, source cancellation, an unlocked source, and no pending timers.

**Comment by luvs01** [comment]
Merged the latest dev (6d19a07) into the PR branch — zero conflicts; new head ce35cdc9. Local gates pass: un run typecheck, un run structure:check, un run privacy:scan; focused tests un test tests/lib/abort-idle-deadline.test.ts tests/server/terminal-guard-server.test.ts tests/responses/responses-core-modules.test.ts are 61 pass / 0 fail. Fork CI dispatched and in progress: diagnostic run 35218067709 (contributor fork Actions)

**Comment by luvs01** [comment]
Absorbed latest dev (`9052ddf7`) — clean merge, new head `eef5364ff` (tree `4569a2ab3`). Gates pass on the merged head (typecheck, structure, privacy) and the three focused test files are 61 pass / 0 fail. Fork CI: diagnostic run 35238601699 (contributor fork Actions) (in progress; the `macos control` dispatch job is expected to cancel at its 30-minute cap — see #4905). Checklist is fully ticked and the PR is out of draft.

**Comment by lidge-jun** [comment]
`macos 1/2` reported `cancelled`, and the aggregate correctly refused to call that a pass — `platform-macos was requested by pull_request but reported 'cancelled'`.

It is a budget overrun, not a failure and not a defect in this branch: the job ran 09:47:34 to 10:07:50, which is 20m16s against the 20-minute `platform-macos` budget. The cause is contention I created by approving nine Cross-platform runs at once; this account runs one of them at a time, and the same shard completed inside the budget earlier today on `dev` and on #4926 and #4938.

Re-running the failed job rather than widening the budget. Widening it would buy green by agreeing to stop noticing a shard that is genuinely wedged, which is the distinction #2152 and #4956 exist to keep.

**Comment by lidge-jun** [comment]
`macos 1/2` failed on a test that is not in this branch's scope, and I do not think it is this branch's defect:

```
(fail) Codex history injection preflight on a WAL store with no live writer >
       reaches a verdict on a cleanly-closed WAL store rather than the catch-all refusal
Received: "history_injection_preflight_unavailable"
```

That case is `tests/codex-integration/codex-history-provider.test.ts:1676`, and this branch touches response body inactivity and adapter continuation/delivery — nothing that reaches SQLite or the history preflight.

Two mundane explanations are ruled out by reading the file. Test isolation holds: the file-level `afterEach` calls `setStateDbPreflightOpenFailureForTests(undefined)`, and this case installs no hook of its own. Sidecar state holds too: `checkpointedWalFixture` removes both `-wal` and `-shm` after closing, which is exactly the condition `openCodexStateForPreflight` requires before admitting the immutable fallback.

What is left is the error classification. The fallback is gated on `isStateDbCantOpenError`, and an error that predicate does not recognise is rethrown into the preflight's catch-all — which is the value observed. #4943 already fought this platform's disagreement about *when* the missing shared memory is raised; this looks like a disagreement about *what* it is raised as, on at least one macOS runner state.

Re-running the failed job. I am tracking it separately rather than treating it as noise: a macOS-only nondeterministic preflight verdict is a flake, and the same file passed on `dev` earlier today in the same shard.

**Comment by lidge-jun** [comment]
Rebased onto current `dev` (`9939b884f0`) and force-pushed. The merge base was `50d72a1d09`, so this branch was testing a merge preview several hours and a dozen merges behind.

Correcting my earlier note: the WAL preflight failure reproduced on the re-run, so it is not a flake in the sense I described. It is deterministic for this branch's merge preview and does not reproduce on `dev`, whose own `macos 1/2` is green at the same time. Staleness is the first thing to rule out, which is what the rebase does.

The second failure in that run, `GET /api/system/memory ... (#848)` at 20249ms, is a timeout under the macOS contention I created by approving nine Cross-platform runs at once, and should clear on its own.

**Comment by lidge-jun** [comment]
Green after the rebase, which settles the diagnosis: the `Codex history injection preflight ... reaches a verdict on a cleanly-closed WAL store` failure was the stale merge preview, not this branch and not the macOS runner. It reproduced twice against a base twelve merges old and does not occur against current `dev`.

Merging. Bounding body inactivity without tearing down the shared abort is the right shape — a stalled body is a property of one response, and cancelling the request's shared signal for it takes down work that is still healthy. Mapping the initial stall to a 504 rather than a generic failure matters for the same reason the budget refusal maps to 429 elsewhere in this tree: the status a client retries on has to match what actually went wrong.
